# Data Wrangling Template

### Introduction 

The Data Quality section focuses on assessing the quality of the data, including any inconsistencies or missing values.
In the WeRateDogs data we will be analysing data by cleaning sections details the steps that need to be taken to clean 
the data, such as removing duplicate records or correcting data values.
furthermore contains the results of data wrangling process and the insights that can be gained from data.

The data we will be looking at is from a twitter account name WeRateDogsates people's dogs with a humorous comment about 
the dog. These ratings almost always have a denominator of 10.Almost always greater than 10. 


### Table of Content
<ol>
    <li><a href="#gather">Gather</a></li>
<li><a href="#assess">Assess data</a>
    <ol>
        <li><a href="#enhanced">Enhanced data frame</a></li> 
        <li><a href="#image">Image dataframe</a></li>
        <li><a href="#tweet">tweet.jason.txt dataframe</a></li>
        <li><a href="#quality">Quality assessment</a></li>
    </ol>
 </li>
    

<li><a href="#making">Making a copy of the data</a>
    <ol>
        <li><a href="#cleaning">Cleaning</a></li>
        <li><a href="#merging">Merging all three dataframe</a></li> 
    </ol>
</li>

<li><a href="#store">Store data</a></li>
<li><a href="#analyzing">Analyzing and visualisation</a></li>
</ol>
    

<a id="gather"></a>
## Gather

In [116]:
# Importing data 
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import requests
import io, json
from decouple import config
import tweepy

In [117]:
# Instantiate variables and objects

# Get env variables stored in a .env file
API_KEY = config("API_KEY", default='')
SECRET_KEY = config("SECRET_KEY", default='')

auth = tweepy.OAuth1UserHandler(API_KEY, SECRET_KEY)
api = tweepy.API(auth, wait_on_rate_limit=True)

In [118]:
# Import The WeRateDogs archive data
enhanced_df = pd.read_csv('twitter-archive-enhanced.csv')

# Get image predictions data and store it in a dataframe
data = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv').text
buffer = io.StringIO(data)
image_df = pd.read_csv(buffer,sep="\t")

## Query the twitter api using enhanced df to create tweet-json.txt

In [119]:
# # NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# # NOTE TO REVIEWER: this student had mobile verification issues so the following
# # Twitter API code was sent to this student from a Udacity instructor
# # Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = enhanced_df.tweet_id.values
len(tweet_ids)

# # Query Twitter's API for JSON data for each tweet ID in the Twitter archive
# count = 0
# fails_dict = {}
# start = timer()
# # Save each tweet's returned JSON as a new line in a .txt file
# with open('tweet_json.txt', 'w') as outfile:
#     # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
#     for tweet_id in tweet_ids:
#         count += 1
#         print(str(count) + ": " + str(tweet_id))
#         try:
#             tweet = api.get_status(tweet_id, tweet_mode='extended')
#             print("Success")
#             json.dump(tweet._json, outfile)
#             outfile.write('\n')
#         except tweepy.TweepError as e:
#             print("Fail")
#             fails_dict[tweet_id] = e
#             pass
# end = timer()
# print(end - start)
# print(fails_dict)


2356

### Read tweet-json.txt to create a dataframe

In [120]:
with open("tweet-json.txt", "r") as file:
    rows = file.read().split("\n")
    data = []

    for row in rows:
        if not row:
            continue
        data.append(json.loads(row))

df = pd.DataFrame(data)



<a id="assess"></a>
## Assess Data

### In section the following will be done 
####  * Use both visual  and programatic assessment.
####  * I will iclude at least eight(8) data quality issues.
####  * Will include at least two(2) tidiness issues.
#### * Document a few issue in a few sentences.

<a id="enhanced"></a>
### Enhanced Twitter Archive

In [121]:
# Displaying data to be analysed 
# there is alot of missing data 
# description headers are too long and close to each other making it difficult to tell apart which header belongs 
# Names sometime abbreviated like a Instead of full sentence 

enhanced_df


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [122]:
# The data seems to be consistent no changes 
enhanced_df.tail()


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [123]:
# looking at the consistancy of the data
enhanced_df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [124]:
enhanced_df.sample()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
459  817827839487737858                    NaN                  NaN   

                     timestamp  \
459  2017-01-07 20:18:46 +0000   

                                                source  \
459  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
459  This is Buddy. He ran into a glass door once. ...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
459                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
459  https://twitter.com/dog_rates/status/817827839...                13   

     rating_denominator   name doggo floofer pupper puppo  
459                  10  Buddy  None    None   None  None

### .info is a compendious summary of the dataframe
 >It shows total coulmns which is 17.
 > Data type 
 > The number of rows and coulumns 
 > The headers of each column 
 > memory usage 

##### Quality assessment -
missing values in_reply_to_status_id,in_reply_to_status_id,retweeted_status_id,retweeted_status_user_id,retweeted_status_timestamp,expanded_urls the total which is showing \
from tweet id and most of the data show a total of 2356 records meaning anything less is missing data.

In [125]:
enhanced_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [126]:
enhanced_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [127]:
# value count 
# enhanced['enhanced'].value_counts() there is an error message 

enhanced_df.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [128]:
enhanced_df.floofer.value_counts()


None       2346
floofer      10
Name: floofer, dtype: int64

In [129]:
enhanced_df.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [130]:
enhanced_df.puppo.value_counts()


None     2326
puppo      30
Name: puppo, dtype: int64

enhanced_df.rating_numerator.sort_values()

In [131]:
enhanced_df.nunique()

tweet_id                      2356
in_reply_to_status_id           77
in_reply_to_user_id             31
timestamp                     2356
source                           4
text                          2356
retweeted_status_id            181
retweeted_status_user_id        25
retweeted_status_timestamp     181
expanded_urls                 2218
rating_numerator                40
rating_denominator              18
name                           957
doggo                            2
floofer                          2
pupper                           2
puppo                            2
dtype: int64

In [132]:
sum(enhanced_df.duplicated())

0

In [133]:
### Total missing data
### Quality assessment- Missing data in in_reply_to_status_id,in_reply_to_user_id,in_reply_to_user_id,retweeted_status_id, retweeted_status_user_id,retweeted_status_user_id, retweeted_status_timestamp, expanded_urls.      
enhanced_df.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

<a id="image"></a>
### 2. Dataframe (Image)

In [134]:
image_df.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [135]:
image_df.info()

# .info gets information of the data including data type 
# The number of rows are not missing.
# It has a total of 12 collumns and 2075 rows.
# Different data type.
# Visualisation assessment -Number of rows are not missing 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [136]:
image_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [137]:
sum(image_df.duplicated())

0

In [138]:
#### Total missing values 
#### There is no missing data in this dataframe 
image_df.isnull().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

<a id="tweet"></a>
### 3. Dataframe (tweet-json.tx)
#### Returning data for tweet jason for analysis 


In [139]:
# Quality assessment- Some rows are missing 
# Dataframe has 31 colunms and 2353 rows
# Different data types
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   created_at                     2354 non-null   object 
 1   id                             2354 non-null   int64  
 2   id_str                         2354 non-null   object 
 3   full_text                      2354 non-null   object 
 4   truncated                      2354 non-null   bool   
 5   display_text_range             2354 non-null   object 
 6   entities                       2354 non-null   object 
 7   extended_entities              2073 non-null   object 
 8   source                         2354 non-null   object 
 9   in_reply_to_status_id          78 non-null     float64
 10  in_reply_to_status_id_str      78 non-null     object 
 11  in_reply_to_user_id            78 non-null     float64
 12  in_reply_to_user_id_str        78 non-null     o

In [140]:
# 
df

created_at                  id              id_str  \
0     Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1     Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2     Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3     Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4     Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   
...                              ...                 ...                 ...   
2349  Mon Nov 16 00:24:50 +0000 2015  666049248165822465  666049248165822465   
2350  Mon Nov 16 00:04:52 +0000 2015  666044226329800704  666044226329800704   
2351  Sun Nov 15 23:21:54 +0000 2015  666033412701032449  666033412701032449   
2352  Sun Nov 15 23:05:30 +0000 2015  666029285002620928  666029285002620928   
2353  Sun Nov 15 22:32:08 +0000 2015  666020888022790149  666020888022790149   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2349  Here we have a 1949 1st generation vulpix. Enj...      False   
2350  This is a purebred Piers Morgan. Loves to Netf...      False   
2351  Here is a very happy pup. Big fan of well-main...      False   
2352  This is a western brown Mitsubishi terrier. Up...      False   
2353  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2349           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2349  {'media': [{'id': 666049244999131136, 'id_str'...   
2350  {'media': [{'id': 666044217047650304, 'id_str'...   
2351  {'media': [{'id': 666033409081393153, 'id_str'...   
2352  {'media': [{'id': 666029276303482880, 'id_str'...   
2353  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/ipho

In [141]:
df.describe()

id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.354000e+03           7.800000e+01         7.800000e+01   
mean   7.426978e+17           7.455079e+17         2.014171e+16   
std    6.852812e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783975e+17           6.757419e+17         3.086374e+08   
50%    7.194596e+17           7.038708e+17         4.196984e+09   
75%    7.993058e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweet_count  favorite_count  quoted_status_id  
count    2354.000000     2354.000000      2.900000e+01  
mean     3164.797366     8080.968564      8.162686e+17  
std      5284.770364    11814.771334      6.164161e+16  
min         0.000000        0.000000      6.721083e+17  
25%       624.500000     1415.000000      7.888183e+17  
50%      1473.500000     3603.500000      8.340867e+17  
75%      3652.000000    10122.250000      8.664587e+17  
max     79515.000000   132810.000000      8.860534e+17

In [142]:
### Total missing data. 
#### In the below columns there is missing data. Quality issue. 
df.isnull().sum()

created_at                          0
id                                  0
id_str                              0
full_text                           0
truncated                           0
display_text_range                  0
entities                            0
extended_entities                 281
source                              0
in_reply_to_status_id            2276
in_reply_to_status_id_str        2276
in_reply_to_user_id              2276
in_reply_to_user_id_str          2276
in_reply_to_screen_name          2276
user                                0
geo                              2354
coordinates                      2354
place                            2353
contributors                     2354
is_quote_status                     0
retweet_count                       0
favorite_count                      0
favorited                           0
retweeted                           0
possibly_sensitive                143
possibly_sensitive_appealable     143
lang        

In [143]:
### Data type 
for col in df.columns:
    if df[col].dtype == object:
        df[col] = df[col].apply(str)
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
2349    False
2350    False
2351    False
2352    False
2353    False
Length: 2354, dtype: bool

<a id="quality"></a>
## Quality assesment 

* Missing values in columns  
* Dog names are sometimes abbreviated like a Instead of full sentence 
* Correct columns with wrong data types 
* Some data in ratting was not extracted correctly 
* The dog names are sometimes first letter capital but sometimes not. Keep the name format consistent.
* The columns’ names are not clear and straightforward such as p1, p2.
* Remove all the unnecessary columns directly ('retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp', 'in_reply_to_status_id', 'in_reply_to_user_id', 'in_reply_to_user_id).
* What need to be done is to capitalize the first letter of dog name for consistence.
* I need to capitalize the first letter of first prediction in image_predictions (I could do that for all the predictions, but I decide to only apply to the first prediction since this variable is the important one).
* The datatype of "timestamp" is not correct.



## Tidiness

 *  The dog stage is the same variable however it is separated in four columns doggo, floofer, pupper, puppo. 
 * Tweet data is spread across the three different/dataframes. 

### Visualisation assessment 

* Missing data. 

<a id="making"></a>
# Making a copy of the data

<a id="clean"></a>
## Clean

In this section 
 * Make a copy of the data before cleaning. 
 * Define-code-test framework.
 * Use and Define-code-test framework.
 * Clean all issues identified in the assessing phase.
 * Create a tidy master dataset with all pieces, cleanied data.

In [144]:
# create a copy of the dataframe before editing 
df_copy = df.copy()
enhanced_df_copy = enhanced_df.copy()
image_df_copy = image_df.copy()

In [145]:
df_copy.head()

created_at                  id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1  Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2  Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3  Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4  Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   
1  This is Tilly. She's just checking pup on you....      False   
2  This is Archie. He is a rare Norwegian Pouncin...      False   
3  This is Darla. She commenced a snooze mid meal...      False   
4  This is Franklin. He would like you to stop ca...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1           [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3            [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4           [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   
2  {'media': [{'id': 891815175371796480, 'id_str'...   
3  {'media': [{'id': 891689552724799489, 'id_str'...   
4  {'media': [{'id': 891327551943041024, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   
1  <a href="http://twitter.com/download/iphone" r...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...                    NaN   
3  <a href="http://twitter.com/download/iphone" r...                    NaN   
4  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ... favorite_count  favorited retweeted possibly_sensitive  \
0  ...          39467      False     False              False   
1  ...          33819      False     False              False   
2  ...          25461      False     False              False   
3  ...          42908      False     False              False   
4  ...          41048      False     False              False   

  possibly_sensitive_appealable lang retweeted_status quoted_status_id  \
0                         False   en              nan              NaN   
1                         False   en              nan              NaN   
2                         False   en              nan              NaN   
3                         False   en              nan              NaN   
4                         False   en              nan              NaN   

  quoted_status_id_str  quoted_status  
0                  nan            nan  
1                  nan            nan  
2                  nan            nan  
3                  nan            nan  
4                  nan            nan  

[5 rows x 31 columns]

In [146]:
enhanced_df_copy.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [147]:
image_df_copy.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Tidiness issue- merging dog stages doggo, floofer, pupper, puppo.

### Define


####  Join dog stage columns  into one called stage 

### Code

In [148]:
# First replace None in stage columns with empty string as follows.

# This do replacements for all four stages

stage = ['doggo','pupper', 'floofer', 'puppo']
  

enhanced_df_copy = enhanced_df_copy.replace('None', '')  


In [149]:
# Combining stage columns 
enhanced_df_copy['stage'] = enhanced_df_copy.doggo.str.cat(enhanced_df_copy.floofer).str.cat(enhanced_df_copy.pupper).str.cat(enhanced_df_copy.puppo)

In [150]:
#  Test
enhanced_df_copy

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

### Dropping columns in enhanced_df_copy

In [151]:
# drop the four old colomns
enhanced_df_copy = enhanced_df_copy.drop(['doggo','floofer','pupper','puppo'], axis = 1)



In [152]:
# Will be using np.na to fill the empty columns 
enhanced_df_copy['stage'] = enhanced_df_copy['stage'].replace('', np.nan)


In [153]:
enhanced_df_copy.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
585   800018252395122689                    NaN                  NaN   
1894  674743008475090944                    NaN                  NaN   
316   834931633769889797                    NaN                  NaN   
1510  691444869282295808                    NaN                  NaN   
1962  673355879178194945                    NaN                  NaN   

                      timestamp  \
585   2016-11-19 16:49:49 +0000   
1894  2015-12-10 00:10:43 +0000   
316   2017-02-24 01:03:08 +0000   
1510  2016-01-25 02:17:57 +0000   
1962  2015-12-06 04:18:46 +0000   

                                                 source  \
585   <a href="http://twitter.com/download/iphone" r...   
1894  <a href="http://twitter.com/download/iphone" r...   
316   <a href="http://twitter.com/download/iphone" r...   
1510  <a href="http://twitter.com/download/iphone" r...   
1962  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
585   Here's a doggo doin a struggle. 11/10 much det...                  NaN   
1894  This is Sophie. She just saw a spider. 10/10 d...                  NaN   
316   This is Tucker. He decided it was time to part...                  NaN   
1510  This is Bailey. She likes flowers. 12/10 https...                  NaN   
1962  This is Koda. She's a boss. Helps shift gears....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
585                        NaN                        NaN   
1894                       NaN                        NaN   
316                        NaN                        NaN   
1510                       NaN                        NaN   
1962                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
585   https://twitter.com/dog_rates/status/800018252...                11   
1894  https://twitter.com/dog_rates/status/674743008...                10   
316   https://twitter.com/dog_rates/status/834931633...                12   
1510  https://twitter.com/dog_rates/status/691444869...                12   
1962  https://twitter.com/dog_rates/status/673355879...                11   

      rating_denominator    name  stage  
585                   10          doggo  
1894                  10  Sophie    NaN  
316                   10  Tucker    NaN  
1510                  10  Bailey    NaN  
1962                  10    Koda    NaN

### define 


Merging df data into enhanced_df using inner join 

In [154]:
# code 

# rename the 'id' from `tweet_data` in preparing for table join
df.rename(columns={'id': 'tweet_id'}, inplace = True)

# check the number of foreign key in two tables
print(enhanced_df_copy.tweet_id.count())
print(df.tweet_id.count())


2356
2354


<a id="merging"></a>
## Merging

In [155]:
# join two tables on 'tweet_id' and use inner join method
# df['tweet_id'].dtype == enhanced_df_copy['tweet_id'].dtype

enhanced_df_copy = pd.merge(enhanced_df_copy, df, on='tweet_id', how = 'inner')

In [156]:
# Test 

enhanced_df_copy.tweet_id.count()

2354

In [157]:
# drop the duplicates due to join
enhanced_df_copy = enhanced_df_copy.drop_duplicates()

In [158]:
print(enhanced_df_copy.tweet_id.count())
print(df.tweet_id.count())

2354
2354


In [159]:
enhanced_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2354 entries, 0 to 2353
Data columns (total 44 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   tweet_id                       2354 non-null   int64  
 1   in_reply_to_status_id_x        78 non-null     float64
 2   in_reply_to_user_id_x          78 non-null     float64
 3   timestamp                      2354 non-null   object 
 4   source_x                       2354 non-null   object 
 5   text                           2354 non-null   object 
 6   retweeted_status_id            179 non-null    float64
 7   retweeted_status_user_id       179 non-null    float64
 8   retweeted_status_timestamp     179 non-null    object 
 9   expanded_urls                  2295 non-null   object 
 10  rating_numerator               2354 non-null   int64  
 11  rating_denominator             2354 non-null   int64  
 12  name                           2354 non-null   o

## Changing wrong datatypes 

### Define

#### Fixing the datatype of the coulmns which have wrong datatypes 

### Code

In [160]:
enhanced_df_copy['timestamp'] = pd.to_datetime(enhanced_df_copy['timestamp'])
enhanced_df_copy['tweet_id'] = enhanced_df_copy['tweet_id'].astype(str)

### Test

In [161]:
print(enhanced_df_copy.timestamp.dtype)
enhanced_df_copy['tweet_id'] = enhanced_df_copy['tweet_id'].astype(str)

datetime64[ns, UTC]


In [162]:
enhanced_df_copy['timestamp'].head(5)

0   2017-08-01 16:23:56+00:00
1   2017-08-01 00:17:27+00:00
2   2017-07-31 00:18:03+00:00
3   2017-07-30 15:58:51+00:00
4   2017-07-29 16:00:24+00:00
Name: timestamp, dtype: datetime64[ns, UTC]

## Missing values 

 ### Define 

#### Replacing missing values with 0 

### Code 

In [163]:
 df.fillna(0, inplace=True)

### Test 

In [164]:
# No missing values 
# filling in missing values with 0

df.isnull().sum()

created_at                       0
tweet_id                         0
id_str                           0
full_text                        0
truncated                        0
display_text_range               0
entities                         0
extended_entities                0
source                           0
in_reply_to_status_id            0
in_reply_to_status_id_str        0
in_reply_to_user_id              0
in_reply_to_user_id_str          0
in_reply_to_screen_name          0
user                             0
geo                              0
coordinates                      0
place                            0
contributors                     0
is_quote_status                  0
retweet_count                    0
favorite_count                   0
favorited                        0
retweeted                        0
possibly_sensitive               0
possibly_sensitive_appealable    0
lang                             0
retweeted_status                 0
quoted_status_id    

## Original retweets data 

### Define 

#### In the guideline of the project it was mention that only analyzation of original tweets
#### We want original ratings 


### Code

In [165]:
enhanced_df_copy = enhanced_df_copy[enhanced_df_copy.retweeted_status_id.isnull()]
enhanced_df_copy = enhanced_df_copy[enhanced_df_copy.retweeted_status_user_id.isnull()]
enhanced_df_copy = enhanced_df_copy[enhanced_df_copy.retweeted_status_timestamp.isnull()]

### Test

In [166]:
enhanced_df_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2353
Data columns (total 44 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   tweet_id                       2175 non-null   object             
 1   in_reply_to_status_id_x        78 non-null     float64            
 2   in_reply_to_user_id_x          78 non-null     float64            
 3   timestamp                      2175 non-null   datetime64[ns, UTC]
 4   source_x                       2175 non-null   object             
 5   text                           2175 non-null   object             
 6   retweeted_status_id            0 non-null      float64            
 7   retweeted_status_user_id       0 non-null      float64            
 8   retweeted_status_timestamp     0 non-null      object             
 9   expanded_urls                  2117 non-null   object             
 10  rating_numerator        

The following columns retweeted_status_id, retweeted_status_id and retweeted_status_timestamp are showing 0 to show that the columns have been removed.

### Some data was not extracted correctly in rating and has data type issue

### Define 

#### Extract data correctly and fix datatype 

### Code

In [167]:
rating = enhanced_df_copy.text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True)
rating.columns = ['rating_numerator', 'rating_denominator']
enhanced_df_copy['rating_numerator'] = rating['rating_numerator'].astype(float)
enhanced_df_copy['rating_denominator'] = rating['rating_denominator'].astype(float)

### Test 

In [168]:
print(enhanced_df_copy.rating_numerator.dtype)
print(enhanced_df_copy.rating_denominator.dtype)

float64
float64


In [169]:
retweeted_status_id.isnull()]
enhanced_df_copy = enhanced_df_copy[enhanced_df_copy.retweeted_status_user_id.isnull()]
enhanced_df_copy = enhanced_df_copy[enhanced_df_copy.retweeted_status_timestamp.isnull()

SyntaxError: unmatched ']' (3466291299.py, line 1)

#### Column called name there are incorrect dog names  e

### Define 

In [ ]:
# Name column has invalid names for example  a.Names start with a capital letter, so the strings in lower case are invalid.

### Code

In [ ]:
for index, row in combined_df.iterrows():
    if row['name'].islower() or row['name'] == "None":
        combined_df.drop(index, inplace= True)

### Test 

In [ ]:
enhanced_df_copy 

### Dropping unnecessary columns  


### Define 

Dropping columns which will not be needed when analysing 

### Code

In [ ]:
# dropping irrelevant columns
enhanced_df_copy.columns
columns_drop = ['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp']

enhanced_df_copy = enhanced_df_copy.drop(columns_drop, axis = 1)

### Test

In [ ]:
enhanced_df_copy.info()

### Define

Capitalize the first letter of dog name for consistence.

### Code 

In [ ]:

enhanced_df_copy['name'].str.islower().sum()

In [ ]:
enhanced_df_copy['name'] = enhanced_df_copy.name.str.capitalize()

### Test

In [ ]:
enhanced_df_copy['name'].str.islower().sum()

### Define 

Change incorrect dog name to none 

### Code 

In [ ]:
# it shows that  it has lots of missing values in name and "a" is even not a name

enhanced_df_copy['name'].value_counts().head(10)

In [ ]:
# choose the most frequent incorrect dog name and set them as None
not_name_list = ['A','The','An']
for i in not_name_list:
        enhanced_df_copy['name'].replace(i, 'None', inplace=True)

### Test 

enhanced_df_copy['name'].value_counts().head(10)

### Define 

### Rename columns for better readabiliy

### Code

In [ ]:
# change the column names

image_df_copy.rename(columns={'p1':'p1_conf', 'first_prediction': 'first_confidence', 'p1_dog': 'first_dog',
                                  'p2': 'second_prediction', 'p2_conf': 'second_confidence', 'p2_dog': 'second_dog',
                                  'p3': 'third_prediction', 'p3_conf': 'third_confidence', 'p3_dog': 'third_dog'}, inplace = True)

### Test 

In [ ]:
image_df_copy

<a id="store"></a>
### Store data 

In [ ]:
enhanced_df_copy.to_csv("twitter_archive_master.csv")

In [ ]:
image_df_copy = pd.read_csv("twitter_archive_master.csv")

<a id="analyzing"></a>
### Analyzing and visualizing data 

#### In this section, analyze and visualize your wrangled data. You must produce at least three (3) insights and one (1) visualization.

Correlation between  retweet_count and favorite_count

In [ ]:
# Sort the DataFrame by "favourites_count" in descending order
df = df.sort_values("favorite_count", ascending=False)

# Select the top 20 rows
top_20 = df.head(20)

# Create a scatter plot of the "retweet_count" and "favourites_count" columns
plt.scatter(top_20["retweet_count"], top_20["favorite_count"])

# Add labels to the x and y axis
plt.xlabel("Retweet Count")
plt.ylabel("favorite Count")

# Show the plot
plt.show()

plt.show()

#### The scatter plot show a positive correlation between favourite_count and retweet_count which could mean that the more a tweet is liked the more chances it has of it being retweeted.


Counted favorite and a possible sensitive appealable 

In [ ]:
df = df.sort_values("favorite_count", ascending=False)
top_10 = df.head(10)
plt.bar(np.arange(10), top_10["favorite_count"])
plt.xticks(np.arange(10), top_10["tweet_id"], rotation=90)
plt.xlabel("favorite_count")
plt.ylabel("possibly_sensitive_appealable") 
plt.show()

The above data was done to anaylyse if one of the reasons why an appeal or a possible appeal might have happened was based on 
the favorite tweet count. It seems like the more the tweeet has been counted as favorite the better chance it has for an appeal of 
the votes done. 


### Different dog stages 

In [ ]:
sorted_stage = enhanced_df_copy['stage'].value_counts().head(3).index
sns.set(style="darkgrid")
sns.countplot(data = enhanced_df_copy, x = 'stage', order = sorted_stage, orient = 'h')
plt.xticks(rotation = 360)
plt.xlabel('Count', fontsize=14)
plt.ylabel('Dog stages', fontsize=14)
plt.title('Different dog stages',fontsize=16)

The division of dog stages.

The stage similarly, I check the division of dog stages. It shows that ‘pupper’ (a small doggo, usually younger) is the most popular dog stage, followed by ‘doggo’ and ‘puppo’. It could be due to the young and unmatured dog is usually cuter than the adult dog. It should also be noticed that there’s huge amount missing data in dog stages, thus the distribution may not reflect the truth.

In [ ]:
image_df_copy['second_prediction'].value_counts().head(10)


#####  The dog breeds above is the top 10 breed this model predicted. The Labrador_retriever is the second  one. It could be because those are the common breeds in U.S. There are more image data on those brees and hence better results. 

Conclusion 

In this project I have explored and cleaned the Tasks in the 
WeRateDogs Twitter data. This data was provided by Udacity and 
required cleaning and wrangling using several different tools and 
techniques. I was able to identify and clean any issues with the 
data and in the end create several 
visualizations to show the distributions of certain variables.